### PR TITLE
 daemon: Add /tofqdns/{id} API

### PR DESCRIPTION
**Pull request by raybejjani**
_Per-commit review is recommended, since I moved some code._

    daemon: Add /discover/fqdn/{id} API

    Introspecting the internal caches used to populate DNS data is useful
    when debugging FQDN policy behaviour and to simply audit what an
    endpoint has done.
    The API returns the entries for each endpoint, instead of using the
    shared global cache. This exposes more per-endpoint history, because the
    global cache only uses the most recent entry from any endpoint. Results
    can be filtered by DNS name, following the policy matchPattern spec, and
    by CIDR range.

```
$ curl -s --unix-socket /var/run/cilium/cilium.sock http://localhost/v1/tofqdns?matchpattern=*.*oogle.com. | jq .
[
  {
    "endpoint-id": 2395,
    "expiration-time": "2019-01-15T16:27:40.395Z",
    "fqdn": "www.google.com.",
    "ips": [
      "74.125.206.105",
      "74.125.206.103",
      "74.125.206.106",
      "74.125.206.147",
      "74.125.206.104",
      "74.125.206.99"
    ],
    "lookup-time": "2019-01-15T15:27:40.395Z",
    "ttl": 3600
  },
  {
    "endpoint-id": 2395,
    "expiration-time": "2019-01-15T16:24:27.835Z",
    "fqdn": "www.google.com.",
    "ips": [
      "2a00:1450:400c:c04::6a"
    ],
    "lookup-time": "2019-01-15T15:24:27.835Z",
    "ttl": 3600
  },
...
 }
]
```

The PR also includes some refactors that might be better left out:
- split out FQDN code into daemon/fqdn.go
- matchpattern.Validate returns the internal regex it compiles
- DNSCache exposes the internal objects via a Dump function

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6652)

### CODE REVIEW

**Comment by raybejjani** [comment]
test-me-please

File: api/v1/openapi.yaml [hunk 1]
```diff
@@ -664,6 +664,55 @@ paths:
         '500':
           description: Metrics cannot be retrieved
 
+  "/tofqdns":
```
Comment (tgraf) [inline api/v1/openapi.yaml]: I would call this `/fqdncache` or `/fqdn/cache`

**Review by tgraf — CHANGES_REQUESTED** [review]
Looks great overall. The matching may be relatively slow so we have to make sure that the API does not get abused

**Comment by raybejjani** [comment]
@tgraf do you mean the nested iterations will be slow? or something else?

Edit: I added a benchmark:
DaemonConsulSuite.BenchmarkFqdnCache        500           3877082 ns/op

Each op is an endpoint with 10000 DNS lookups, each with 10 IPs. The matchpattern and CIDR values are match-all, but not the default `return true` (so they should behave closer to the worst case).

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Pull request merged**